swarm/api/config.go

```diff
@@ -46,101 +43,63 @@ type Config struct {
-	address := crypto.PubkeyToAddress(prvKey.PublicKey) // default beneficiary address
-	dirpath := filepath.Join(path, "bzz-"+common.Bytes2Hex(address.Bytes()))
-	err = os.MkdirAll(dirpath, os.ModePerm)
-	if err != nil {
-		return
-	}
-	confpath := filepath.Join(dirpath, "config.json")
-	var data []byte
-	pubkey := crypto.FromECDSAPub(&prvKey.PublicKey)
-	pubkeyhex := common.ToHex(pubkey)
-	keyhex := crypto.Keccak256Hash(pubkey).Hex()
+//create a default config with all parameters to set to defaults
+func NewDefaultConfig() (self *Config) {
 
 	self = &Config{
-		SyncParams:    network.NewSyncParams(dirpath),
-		HiveParams:    network.NewHiveParams(dirpath),
+		StoreParams:   storage.NewDefaultStoreParams(),
 		ChunkerParams: storage.NewChunkerParams(),
-		StoreParams:   storage.NewStoreParams(dirpath),
+		HiveParams:    network.NewDefaultHiveParams(),
+		SyncParams:    network.NewDefaultSyncParams(),
+		Swap:          swap.NewDefaultSwapParams(),
 		ListenAddr:    DefaultHTTPListenAddr,
 		Port:          DefaultHTTPPort,
-		Path:          dirpath,
-		Swap:          swap.DefaultSwapParams(contract, prvKey),
-		PublicKey:     pubkeyhex,
-		BzzKey:        keyhex,
+		Path:          node.DefaultDataDir(),
+		EnsApi:        node.DefaultIPCEndpoint("geth"),
 		EnsRoot:       ens.TestNetAddress,
-		NetworkId:     networkId,
-	}
-	data, err = ioutil.ReadFile(confpath)
-
-	// if not set in function param, then set default for swarm network, will be overwritten by config file if present
-	if networkId == 0 {
-		self.NetworkId = network.NetworkId
+		NetworkId:     network.NetworkId,
+		SwapEnabled:   false,
+		SyncEnabled:   true,
+		SwapApi:       "",
 	}
 
-	if err != nil {
-		if !os.IsNotExist(err) {
-			return
-		}
-
-		// file does not exist
-		// write out config file
-		err = self.Save()
-		if err != nil {
-			err = fmt.Errorf("error writing config: %v", err)
-		}
-		return
-	}
+	return
+}
 
-	// file exists, deserialise
-	err = json.Unmarshal(data, self)
-	if err != nil {
-		return nil, fmt.Errorf("unable to parse config: %v", err)
-	}
-	// check public key
-	if pubkeyhex != self.PublicKey {
-		return nil, fmt.Errorf("public key does not match the one in the config file %v != %v", pubkeyhex, self.PublicKey)
-	}
-	if keyhex != self.BzzKey {
-		return nil, fmt.Errorf("bzz key does not match the one in the config file %v != %v", keyhex, self.BzzKey)
-	}
+//some config params need to be initialized after the complete
+//config building phase is completed (e.g. due to overriding flags)
+func (self *Config) Init(prvKey *ecdsa.PrivateKey) {
 
-	// if set in function param, replace id set from config file
-	if networkId != 0 {
-		self.NetworkId = networkId
-	}
+	pubkey := crypto.FromECDSAPub(&prvKey.PublicKey)
+	pubkeyhex := common.ToHex(pubkey)
+	keyhex := crypto.Keccak256Hash(pubkey).Hex()
 
-	self.Swap.SetKey(prvKey)
+	self.PublicKey = pubkeyhex
+	self.BzzKey = keyhex
 
```

Comment:
I was under the impression that we did not want to be able to infer the overlay address from predictable underlying data?

Reply:
? 
I should have not changed any logic in this respect, just reorganizing stuff which was there before, but only allowing things to happen after the complete config  with the overrides has been evaluated.

This is the old way, where the config was initialized in `swarm/api/config.go` with `NewConfig()`:

```
60 func NewConfig(path string, contract common.Address, prvKey *ecdsa.PrivateKey, networkId uint64) (self *Config, err error) {
 61   address := crypto.PubkeyToAddress(prvKey.PublicKey) // default beneficiary address
 62   dirpath := filepath.Join(path, "bzz-"+common.Bytes2Hex(address.Bytes()))
 63   err = os.MkdirAll(dirpath, os.ModePerm)
 64   if err != nil {
 65     return
 66   }
 67   confpath := filepath.Join(dirpath, "config.json")
 68   var data []byte
 69   pubkey := crypto.FromECDSAPub(&prvKey.PublicKey)
 70   pubkeyhex := common.ToHex(pubkey)
 71   keyhex := crypto.Keccak256Hash(pubkey).Hex()
 72 
 73   self = &Config{
 74     SyncParams:    network.NewSyncParams(dirpath),
 75     HiveParams:    network.NewHiveParams(dirpath),
 76     ChunkerParams: storage.NewChunkerParams(),
 77     StoreParams:   storage.NewStoreParams(dirpath),
 78     ListenAddr:    DefaultHTTPListenAddr,
 79     Port:          DefaultHTTPPort,
 80     Path:          dirpath,
 81     Swap:          swap.DefaultSwapParams(contract, prvKey),
 82     PublicKey:     pubkeyhex,
 83     BzzKey:        keyhex,
 84     EnsRoot:       ens.TestNetAddress,
 85     NetworkId:     networkId,
 86   }
```

Or I did misunderstand your comment :)

Reply:
No from my understanding of the ultimate intention, the old method is also wrong. I though the overlay address should be totally independent / random. We've discussed this erratically a few times, but never really had a firm discussion about it, which maybe we should. We can address it in a separate issue, so I suggest leave this for now as it is.